firmware/Core/Src/stm32_internal_flash_drivers/stm32_internal_flash_drivers.c

```diff
@@ -0,0 +1,143 @@
+#include "stm32_internal_flash_drivers/stm32_internal_flash_drivers.h"
+
+#include "stm32l4r5xx.h"
+#include "stm32l4xx_hal.h" // must include this before stm32l4xx_hal_flash.h
+#include "stm32l4xx_hal_flash.h"
+#include "stm32l4xx_hal_flash_ex.h"
+#include <string.h>
+
+/// @brief Writes data to the flash memory in chunks of 8 bytes.
+/// @param address Address in the flash memory where the data will be written.
+/// @param data uint8_t buffer containing the data to be written.
+/// @param length Length of the data to be written.
+/// @return 0 on success, > 0 on error, 10 if HAL_FLASH_Unlock() failed, 11 if HAL_FLASH_Lock() failed
+/// @note Currently, only allowed to write to golden copy region
+/// @note if data is not 8 bytes long, it will write the data given, and clear the rest of the data
+/// Ex: write: 0x01020304, will write 0x0102030400000000, clearing the 8 bytes infront
+uint8_t STM32_internal_flash_write(uint32_t address, uint8_t *data, uint32_t length)
+{
+    if (address < STM32_INTERNAL_FLASH_MEMORY_REGION_GOLDEN_COPY_ADDRESS)
+    {
+        return 1;
+    }
+
+    const uint32_t end_address = address + length;
+
+    if (end_address > FLASH_BANK2_END)
+    {
+        return 2;
+    }
+
+    if (HAL_FLASH_Unlock() != HAL_OK)
+    {
+        return 10;
+    }
+
+    // Clear all FLASH flags before starting the operation
+    __HAL_FLASH_CLEAR_FLAG(FLASH_FLAG_ALL_ERRORS);
+
+    HAL_StatusTypeDef status = HAL_OK;
+
+    for (uint32_t currentAddress = address; currentAddress < end_address; currentAddress += 8)
+    {
+        uint8_t data_to_write[8] = {0};
+
+        // TODO: what to do if data is not 8 bytes long
+        // Currently, it will set the rest of the values to 0
+
+        // currentAddress - address is the number of bytes we have written
+        // since the beginning of the function
+        memcpy(data_to_write, data + (currentAddress - address), 8);
+
+        const uint64_t double_word = *(uint64_t *)(data_to_write);
+
+        status = HAL_FLASH_Program(FLASH_TYPEPROGRAM_DOUBLEWORD, currentAddress, double_word);
+        if (status != HAL_OK)
+        {
+            break;
```

Comment:
No change required if you disagree, but I finally figure out why this `break;` bothers me so much (instead of returning from there): You generally want the reported error/exeception to be the first error that happens. 

Consider the case where `HAL_FLASH_Program` fails, and then also fails to lock. The return code will be that it failed to lock, when the useful indicator is actually that it failed `HAL_FLASH_Program` in my opinion, and then subsequently failed to lock.

Reply:
I think that's really fair. 

I have a couple of alternatives:

1. First
    1. That if statement returns value (lets say 3)
    2. The calling function inspects the return value, if it is 3, it will execute the following code:
```
const uint8_t write_res = stm32_internal_flash_write(...);
if (write_res == 3) {
const HAL_StatusTypeDef lock_res = HAL_Flash_Lock();
if (lock_res != HAL_OK) {
    snprintf(....);
    return 1;
    }
snprintf(....);
return 1;
}

```
    3. This is a bit messy but not sure how else to remedy this.
2. Second
    1. The function never locks internally, rather, the calling code always must call HAL_Flash_Lock() after using this function
    2. I personally don't like this because then the API layer is not segregated but then, we don't have diverting paths for how this function Locks the Flash.

Reply:
With both of these suggestions, the API layer will become exposed to the calling code, so I don't know how to avoid that.

1 more thing I was able to come up with is the following: 

- Essentially, we have 4 states, either:
    - Everything is successful
    - Write fails but lock is successful
    - Write is successful but lock fails
    - Write fails and lock fails.

Currently, the approach addresses the first 3, but is missing the last one. I was wondering if you had any suggestions on restructuring the code so that I don't have to expose the internal API layer to the telecommand function

Reply:
Will leave to you to decide! We're not playing code golf though; calling the lock function in two places isn't a sin.

Reply:
Yes, I just realized that I over complicated it.